rbs/CommentsAssociator.cc

```diff
@@ -15,23 +16,95 @@ const string_view CommentsAssociator::RBS_PREFIX = "#:";
 const string_view CommentsAssociator::ANNOTATION_PREFIX = "# @";
 const string_view CommentsAssociator::MULTILINE_RBS_PREFIX = "#|";
 
-namespace {
-bool isVisibilitySend(const parser::Send *send) {
-    return send->receiver == nullptr && send->args.size() == 1 &&
-           (parser::isa_node<parser::DefMethod>(send->args[0].get()) ||
-            parser::isa_node<parser::DefS>(send->args[0].get())) &&
-           (send->method == core::Names::private_() || send->method == core::Names::protected_() ||
-            send->method == core::Names::public_() || send->method == core::Names::privateClassMethod() ||
-            send->method == core::Names::publicClassMethod() || send->method == core::Names::packagePrivate() ||
-            send->method == core::Names::packagePrivateClassMethod());
+// Static regex pattern to avoid recompilation
+static const regex HEREDOC_PATTERN("\\s*=?\\s*<<(-|~)[^,\\s\\n#]+(,\\s*<<(-|~)[^,\\s\\n#]+)*");
+
+/**
+ * Check if the given range is the start of a heredoc assignment `= <<~FOO` and return the position of the end of the
+ * heredoc marker.
+ *
+ * Returns -1 if no heredoc marker is found.
+ */
+uint32_t hasHeredocMarker(core::Context ctx, const uint32_t fromPos, const uint32_t toPos) {
+    string_view source(ctx.file.data(ctx).source().substr(fromPos, toPos - fromPos));
+
+    string source_str(source);
+    smatch match;
+    if (regex_search(source_str, HEREDOC_PATTERN)) {
+        return fromPos + source_str.length();
+    }
+    return UINT32_MAX;
 }
 
-bool isAttrAccessorSend(const parser::Send *send) {
-    return (send->receiver == nullptr || parser::isa_node<parser::Self>(send->receiver.get())) &&
-           (send->method == core::Names::attrReader() || send->method == core::Names::attrWriter() ||
-            send->method == core::Names::attrAccessor());
+uint32_t CommentsAssociator::locateTargetLine(parser::Node *node) {
```

Comment:
What do you think about using an `std::optional` here instead of `UINT32_MAX`? Experimenting on godbolt shows that it will always return through a register when optimizations are enabled.

Reply:
Agreed, done 👍